### PR TITLE
Update Substrate to use latest GraalVM (JDK 23) and JavaFX 24-ea

### LINKED ISSUE #1241: Request for Gluon/Graal release
There was no gluon graalvm release for a year and a half. The last edition is very old. Have you abandoned this project already? If so, is there an alternative? I want to build for Android (a spring boot project), but it is only possible with newer graal.

### PR DESCRIPTION
**Pull request by jperedadnr**
### Issue

Fixes #1241 

### Progress

- [ ] Change must not contain extraneous whitespace
- [ ] License header year is updated, if required
- [ ] Verify the contributor has signed [Gluon Individual Contributor License Agreement (CLA)](https://docs.google.com/forms/d/16aoFTmzs8lZTfiyrEm8YgMqMYaGQl0J8wA0VJE2LCCY)

### CODE REVIEW

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 1]
```diff
@@ -324,7 +317,28 @@ public boolean createSharedLib() throws IOException, InterruptedException {
...
+        if (!link()) {
+            Logger.logSevere("Error building a static library: error linking the native image");
+            return false;
+        }
+        ProcessRunner createStaticLibRunner = new ProcessRunner(getStaticLinker());
+        createStaticLibRunner.addArg(getStaticLinkerArgs());
+        Path dest = paths.getGvmPath().resolve("lib" + projectConfiguration.getAppName() + ".a");
+        createStaticLibRunner.addArg(dest.toString());
+        createStaticLibRunner.addArg(getProjectObjectFile().toString());
+        createStaticLibRunner.addArgs(getAdditionalObjectFiles());
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java:333]: there should be no additional object files

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 2]
```diff
@@ -324,7 +317,28 @@ public boolean createSharedLib() throws IOException, InterruptedException {
...
+        createStaticLibRunner.setInfo(true);
+        createStaticLibRunner.setLogToFile(true);
+        int result = createStaticLibRunner.runProcess("archive");
+        if (result == 0) {
+            // copy vmone lib to target, as it will be needed for linking
+            FileOps.copyDirectory(projectConfiguration.getJavaStaticLibsPath(), paths.getGvmPath());
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java:339]: Probably good to print this out. Developers are going to need this path.

File: src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java [hunk 3]
```diff
@@ -169,7 +165,7 @@ public String getAdditionalSourceFileLocation() {
     @Override
     List<String> getAdditionalSourceFiles() {
         List<String> answer = new ArrayList<>(iosAdditionalDummySourceFiles);
-        if (!projectConfiguration.isSharedLibrary()) {
+        if (!projectConfiguration.isSharedLibrary() && !projectConfiguration.isStaticLibrary()) {
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java:168]: idem

File: src/main/java/com/gluonhq/substrate/target/LinuxTargetConfiguration.java [hunk 4]
```diff
@@ -221,52 +210,20 @@ public boolean runUntilEnd() throws IOException, InterruptedException {
...
     List<String> getOtherStaticLibs() {
-        return Stream.concat(staticJvmLibs.stream().map(lib -> ":lib" + lib + ".a"), linuxLibs.stream())
+        return Stream.concat(getStaticJavaLibs().stream().map(lib -> ":lib" + lib + ".a"), linuxLibs.stream())
                 .collect(Collectors.toList());
     }
 
     @Override
     List<String> getAdditionalSourceFiles() {
-        if (projectConfiguration.isSharedLibrary()) {
+        if (projectConfiguration.isSharedLibrary() || projectConfiguration.isStaticLibrary()) {
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/LinuxTargetConfiguration.java:221]: idem

File: src/main/java/com/gluonhq/substrate/target/WindowsTargetConfiguration.java [hunk 5]
```diff
@@ -337,12 +311,18 @@ private boolean executableOnPath(String executable) {
 
     @Override
     List<String> getAdditionalSourceFiles() {
-        if (projectConfiguration.isSharedLibrary()) {
+        if (projectConfiguration.isSharedLibrary() || projectConfiguration.isStaticLibrary()) {
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/WindowsTargetConfiguration.java:314]: idem

File: src/main/java/com/gluonhq/substrate/util/FileOps.java [hunk 6]
```diff
@@ -738,6 +745,41 @@ private static String generateClasspathFromTemporaryFolder(Path tmpPath, String
...
+        try {
+            Files.createDirectories(filePath.getParent());
+
+            Logger.logDebug("download from link: " + link);
+            HttpRequest request = HttpRequest.newBuilder()
+                    .uri(URI.create(link)).build();
+            HttpResponse<Void> response = HttpClient.newHttpClient()
+                    .send(request, HttpResponse.BodyHandlers.discarding());
+
+            Optional<String> location = response.headers().firstValue("location");
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/util/FileOps.java:764]: best to check 304 status code here.

Comment (johanvos) [reply src/main/java/com/gluonhq/substrate/util/FileOps.java:764]: I meant 302 indeed

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/util/FileOps.java:767]: done

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java (hunk 1 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: removed

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 7]
```diff
@@ -346,7 +331,7 @@ public String getAdditionalSourceFileLocation() {
     @Override
     List<String> getAdditionalSourceFiles() {
         List<String> answer = new ArrayList<>(androidAdditionalDummySourceFiles);
-        if (!projectConfiguration.isSharedLibrary()) {
+        if (!projectConfiguration.isSharedLibrary() && !projectConfiguration.isStaticLibrary()) {
```
[the remark this answers is not in the document]
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: removed

File: src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java (hunk 3 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java]: removed

File: src/main/java/com/gluonhq/substrate/target/LinuxTargetConfiguration.java (hunk 4 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/LinuxTargetConfiguration.java]: removed

File: src/main/java/com/gluonhq/substrate/target/MacOSTargetConfiguration.java [hunk 8]
```diff
@@ -86,7 +77,7 @@ String getAdditionalSourceFileLocation() {
 
     @Override
     List<String> getAdditionalSourceFiles() {
-        if (projectConfiguration.isSharedLibrary()) {
+        if (projectConfiguration.isSharedLibrary() || projectConfiguration.isStaticLibrary()) {
```
[the remark this answers is not in the document]
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/MacOSTargetConfiguration.java]: removed

File: src/main/java/com/gluonhq/substrate/target/WindowsTargetConfiguration.java (hunk 5 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/WindowsTargetConfiguration.java]: removed

**Pull request merged**